Fix documentation 3 (#4798)
widelands/widelands: doc/sphinx/source/tutorial.rst

.. code-block:: lua

   include "scripting/coroutine.lua"

   function print_a_word(word)

>       while 1 do

Reviewer: ```suggestion
      while true do
```

Author: done

         print(word)

>          sleep(1000)         -- always sleep a bit, otherwise the system got stuck

Reviewer: ```suggestion
         sleep(1000)         -- always sleep a bit in loops, otherwise Widelands will hang
```

Author: It's not just widelands... i got sometimes there loosing control over the whole system. No mouse, no taskmanager, not able to switch windows... the only thing i could do was to press the main power button...

Reviewer: I find this hard to believe since operating systems should surely never grant a single process full access to all the CPU power for more than a few milliseconds at a time? #4334 is another such weird thing. I'm not convinced that it is Widelands that effects such a total system freeze; I've seen a fair share of such freezes that seem to coincide suspiciously often with a particular program and yet cannot possibly be related.

I've tried out entering an endless loop in the debug console now, and while Widelands got hung up as expected, the rest of the system kept behaving just fine.

Author: Can't reproduce this in the debug console also. But surely i got there in the past while testing scenarios …

Anyway: The warning should be as hard as it can be, imho. It's both a situation which is bad, not all users know how to kill a process.

Reviewer: Okay. But then how about putting this info prominently in a `.. note::` section instead of mentioning it on the side in a code example? E.g.
```
.. note:: 
   Always add :meth:`sleep` statements to lengthy loops.
   Widelands will hang while the code between two :meth:`sleep`
   calls is executed, resulting in a risk that the program – or
   even the whole operating system – may stall for an extended
   period of time, or even until Widelands is force-closed.
```

Author: Since there was already a not, i wrote a mix of my own words, the already written note and your suggestion.

.. code-block:: lua

   include "scripting/coroutine.lua"

>    function print_a_word(word)         -- this function is a coroutine

Reviewer: Strictly this is not correct (and the paragraph above is also overly simplistic). `print_a_word` is a *function*. The call to `run(f)` then creates a new coroutine object from the function `f`. A coroutine, in Widelands's Lua implementation, is a wrapper *object* around a function in a *Lua State*, with its own *Lua stack* and set of local variables. Most of the time this is a fine distinction, but when you want more control over your script's workflow it becomes a very important one.
We don't need to explain it in such detail in the beginner's tutorial but we shouldn't give incorrect information either.

Author: " -- this function will get a coroutine" ?

Reviewer: "get" might be interpreted as "receive" in this context. It's a word with many different meanings which we're overfond of using when translating German into English ;D

"Function from which a coroutine will later be constructed" or "A function we'll use to create a coroutine" would be clearer imho. 
This could be clarified further by changing the comment on the `run(...)` line to "Constructs a new coroutine from the function `print_a_word` and an argument, and immediately launches this coroutine."

Author: Good explanation :+1: 

Done

.. code-block:: lua

   print pretty_plurals_string:bformat(number_of_worlds)

If you have a string with multiple numbers in it that would trigger plural
forms, split it into separate strings that you can fetch with ``ngettext``.
You can then combine them with ``bformat`` and ordered placeholders.

Handling Long Strings
^^^^^^^^^^^^^^^^^^^^^

If you have a really long string, e.g. a dialog stretching over multiple
sentences, check if there is a logical place where you could split this into
two separate strings for translators. We don't have a "break after x characters"
rule for this; please use common sense here. It is easier for translators to
translate smaller chunks, and if you should have to change the string later on,
e.g. to fix a typo, you will break less translations. The strings will be put
into the translation files in the same order as they appear in the source code,
so the context will remain intact for the translators.

Also, please hide all formatting control characters from our translators. This
includes HTML tags as well as new lines in the code! For an example, have a

> look at ``data/campaigns/atl01.wmf/scripting/texts.lua``.

Reviewer: Also add a note that one must not simply concatenate two sentences but instead use bformat or a localized-concatenation function (don't remember what it's called and where it's defined), unless there's a paragraphdivider or other richtext markup between the sentences.

Author: I have not enough knowledge about this localization stuff to write such a note.

Changed HTML with a link to richtext.lua.

Reviewer: ```
Note that simply concatenating two translatable strings with ``..``
does not work in all languages. To achieve the desired effect, use
the function :meth:`join_sentences` with the two sentences as
arguments, or :meth:`paragraphdivider` if you wish the text to
continue on a new paragraph.

Never assemble a single localized sentence with the ``..``
operator. Use :meth:`string.bformat` with appropriate
placeholder substitutions for such cases.
```

Author: Thanks, done.
